### ISSUE TITLE
Shared registration: NEXT_IDLE invocation_policy 

### ISSUE THREAD

**Issue by konsultaner** [wamp-proto/wamp-proto#364]
@oberstet What do you think about a `invocation_policy` `NEXT_IDLE`? 

Lets say 3 callees have a shared registration on a procedure. The procedure takes quite some time to compute. If callee 1 has not finished processing the invocation and not yet answered with a result, the router must take callee 2. If callee 2 is busy with the same procedure callee3 is chosen.

This would be a good way to kind of autoscale business logic. A router implementation could auto-start more clients if lets say n-1 callees are busy with a chosen procedure, but this is a little off topic.

**Comment by oberstet** [comment]
as far as I see, we already have a feature in WAMP AP that achieves the same effect. that is by setting the "maximum concurrency" when registering a shared procedure:

https://github.com/crossbario/crossbar-examples/tree/master/scaling-microservices#concurrency-control

**Comment by konsultaner** [comment]
@oberstet Thats would be a good thing to have in the specs!

With `NEXT_IDLE` I mean something different. For example `ROUND_ROBIN` always takes the `(n+1) mod m` callee and `RANDOM` a random of the `m` available callees to execute the computation. `NEXT_IDLE` should always take the first one, but only if its not busy, otherwize the second one and so on. If the first callee is idle again, it should take first callee again. 

If you think this is totally off. What do you think of `X_` for custom invocation policy prefix?

**Comment by binaek89** [comment]
@konsultaner as far as I know, as @oberstet mentioned, `shared_registration`s have a concurrency and call-queuing mechanism. Although not mentioned in the `protocol specification`, it exists in the official router implementation.

**Comment by binaek89** [comment]
This along with `registration` `statistics` could be used as a metric for the orchestrator to add or remove applications as required by the actual application load. Much like how HPA works in K8S.

This will probably require extending the `registration` `statistics` data to include some more information like:
* last_called_at
* last_idled_at
* *etc*

**Comment by oberstet** [comment]
> NEXT_IDLE should always take the first one, but only if its not busy, otherwize the second one and so on.

this is how the feature ("call queuing") works in Crossbar.io

**Comment by konsultaner** [comment]
@oberstet I'm basically talking about `concurrency = 1`. I don't need it to specify parallelism. I have tasks like "invoice accounting" or other long running calculations, that should be called only once per callee at a time.

I actually like the `concurrency` option, because you can combine it with either round robin or random. This is more flexible then having 2 more policies. I would love to see this in the specs.

What happens if 2 clients run `concurreny = 2` and the 5th caller wants call the procedure? Will the router cancel the invocation or just queue it up?

> This along with registration statistics could be used as a metric for the orchestrator to add or remove applications as required by the actual application load. Much like how HPA works in K8S.
> 
> This will probably require extending the registration statistics data to include some more information like:
> 
> last_called_at
> last_idled_at
> etc

@binaek89 I would also really like to see this. I'm currently working on autoscaling for my router implementation.

**Comment by oberstet** [comment]
> I'm basically talking about concurrency = 1. I don't need it to specify parallelism. I have tasks like "invoice accounting" or other long running calculations, that should be called only once per callee at a time.

yes, that will work with the concurrency feature. concurrency == 1 effectively means: the callee is full serialized (in the DB sense).

> What happens if 2 clients run concurreny = 2 and the 5th caller wants call the procedure? Will the router cancel the invocation or just queue it up? 

You can control that via router configuration: either (immediately) return an error to the caller OR queue up the call (with a configurable queue depth .. upon hitting that, again returning errors then).

---

@binaek89 I think these could be actually 2 features (well, as far as I understand what you are after):

1. automatic service activation
2. callee invocation statistics

Automatic service activation could make a router automatically _start_ a WAMP application component / service _triggered_ by a publication to a given topic or a call to a given procedure.

It could _also_ automatically start additional instances of a WAMP application component / service when an incoming call cannot be served by any callee due to maximum total concurrency reached, call queue full, ...

**Comment by binaek89** [comment]
I believe automatic service activation/deactivation is beyond the scope of the protocol (except maybe publishing the fact that it's running on overload, as you mentioned).

However, having some extra data in the invocation statistics available will be an invaluable asset for the  infrastructure orchestrator when it needs to scale the infrastructure (more so when it's downscaling and needs to decide which instance to take down).
